src/estimation/tests/KalmanFilterUnitTest.cpp

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (C) 2019 Fondazione Istituto Italiano di Tecnologia
+ *
+ * Licensed under either the GNU Lesser General Public License v3.0 :
+ * https://www.gnu.org/licenses/lgpl-3.0.html
+ * or the GNU Lesser General Public License v2.1 :
+ * https://www.gnu.org/licenses/old-licenses/lgpl-2.1.html
+ * at your option.
+ */
+
+#include <iDynTree/Estimation/KalmanFilter.h>
+#include <iDynTree/Core/TestUtils.h>
+#include <iostream>
+#include <memory>
+
+#include <random>
+
+int main()
+{
+    iDynTree::DiscreteKalmanFilterHelper kf;
+    // state space is position and velocity of truck with constant input acceleration u = a
+    // x {k+1} = A x {k} + B u {k}
+    // x1 {k+1} = x1 {k} + dt x2 {k} + (0.5) (dt^2) u
+    // x2 {k+1} = x2 {k} + dt u
+
+    double dt{1.0}; ///< discretization_time_step_in_s
+    iDynTree::MatrixDynSize A;
+    A.resize(2, 2);
+    A(0, 0) = 1; A(0, 1) = dt;
+    A(1, 0) = 0; A(1, 1);
+
+    iDynTree::MatrixDynSize B;
+    B.resize(2, 1);
+    B(0, 0) = 0.5*dt*dt;
+    B(1, 0) = dt;
+
+    // noisy measurement of the truck is made at each step without any feedthrough
+    iDynTree::MatrixDynSize C, D;
+    C.resize(1, 2);
+    C(0, 0) = 1; C(0, 1) = 2;
+
+    bool ok{false};
+    ok = kf.constructKalmanFilter(A, B, C);
+    ASSERT_IS_TRUE(ok);
+    std::cout << "Kalman filter constructed successfully." << std::endl;
+
+    iDynTree::MatrixDynSize Q;
+    Q.resize(2, 2);
+    double system_noise_var{1.0};
+    iDynTree::toEigen(Q) = system_noise_var*Eigen::MatrixXd::Identity(2, 2);
+    ok = kf.kfSetSystemNoiseCovariance(Q);
+    ASSERT_IS_TRUE(ok);
+    std::cout << "System noise covariance matrix set." << std::endl;
+
+    iDynTree::MatrixDynSize R;
+    R.resize(1, 1);
+    double measurement_noise_var{1.0};
+    R(0, 0) = measurement_noise_var;
+    ok = kf.kfSetMeasurementNoiseCovariance(R);
+    ASSERT_IS_TRUE(ok);
+    std::cout << "Measurement noise covariance matrix set." << std::endl;
+
+    iDynTree::VectorDynSize x0;
+    x0.resize(2);
+    x0.zero();
+    ok = kf.kfSetInitialState(x0);
+    ASSERT_IS_TRUE(ok);
+    std::cout << "initial state set." << std::endl;
+
+    iDynTree::MatrixDynSize P0;
+    P0.resize(2, 2);
+    double state_var{1.0};
+    iDynTree::toEigen(P0) = state_var*Eigen::MatrixXd::Identity(2, 2);
+    ok = kf.kfSetStateCovariance(P0);
+    std::cout << "initial state covariance set." << std::endl;
+
+    ok = kf.kfInit();
+    ASSERT_IS_TRUE(ok);
+    std::cout << "Kalman filter initialized." << std::endl;
+
+    // adding Gaussian noise to measurement
+    const double mean = 0.0;
+    const double stddev = 1.0;
+    std::mt19937 generator(std::random_device{}());
+    auto dist = std::bind(std::normal_distribution<double>{mean, stddev},
+                    std::mt19937(std::random_device{}()));
```

Comment:
I strongly suggest not to use actually random number as seed for the test, as the day the test will actually fail, you will never be able to reproduce the failing test. Just use an arbitrary deterministic seed for the RNG, and this will make the test deterministic.

Reply:
Avoiding non-determinism in tests is also useful if we ever want to do fuzzing, see https://llvm.org/docs/LibFuzzer.html#fuzz-target .

Reply:
Removed `std::random_device` and and used stdlib's srand and rand instead. Fixed in the recent commit.